### DISCUSSION TITLE
[Auto3dSeg] Inference for new cases

### DISCUSSION THREAD

**Discussion by imruljubair** [Project-MONAI/MONAI#7481] [Q&A]
Hi,

Question about: Auto3dSeg.

I used `AutoRunner` for `algo=segresnet`. There were 4 folds. I finished training for all folds and got results in _ensembled_output_. Now I want to run inference for a new set of test files (may be in different directory) with my trained ensemble model.

I believe there is an easy solutions to do that, which I can't find in tutorial. May be I missed something.

Can anyone guide me please?

Jubair

**Comment by KumoLiu** [comment]
Hi @imruljubair, thanks for your interest here.
You can use `EnsembleRunner` directly.
https://github.com/Project-MONAI/MONAI/blob/50f9aea67fb1ea7967020ad613ce83409261f2de/monai/apps/auto3dseg/ensemble_builder.py#L397

Also here is a tutorial about how to customize your ensemble pipeline by defining a new ensemble class.
https://github.com/Project-MONAI/tutorials/blob/main/auto3dseg/notebooks/ensemble_byoc.ipynb

Hope it helps, thanks.

**Reply by imruljubair** [reply]
@KumoLiu 
Thank you.
I need to use this: `ensemble_runner = EnsembleRunner(data_src_cfg_name,
                                             work_dir,
                                             ensemble_method_name,
                                             mgpu=device_setting['n_devices']>1,
                                             **kwargs,
                                             **pred_params)`

Could you please tell me what is `data_src_cfg_name`?

Thanks.

**Reply by KumoLiu** [reply]
It is the str that points to the data source path.
An example here:
https://github.com/Project-MONAI/tutorials/blob/main/auto3dseg/tasks/btcv/input.yaml

**Reply by imruljubair** [reply]
@KumoLiu 
Is it possible to perform predictions on the files which are not added in the `folds.json`. For instances, in my `folds.json` I have -
```
    "testing": [
        {
            "image": "./imagesTs/pat43.nii.gz"
        },
        {
            "image": "./imagesTs/pat57.nii.gz"
        },
        {
            "image": "./imagesTs/pat29.nii.gz"
        },
        {
            "image": "./imagesTs/pat60.nii.gz"
        }
               ]
```

But I have new test cases which are in different directory and different names. Can I do that if I do not want to regenerate the `folds.json`?

**Reply by KumoLiu** [reply]
Hi @imruljubair,
Currently, the functionality you're interested in doesn't have a straightforward implementation without making some alterations to the code. If you would wish to make these modifications, refer to this section of the codebase: 
https://github.com/Project-MONAI/MONAI/blob/f6f9e817892090f6c1647f60f51fbc5ed80b74f1/monai/apps/auto3dseg/ensemble_builder.py#L387
Additionally, we have a tutorial on preparing the data, which could prove beneficial for you: [MSD Datalist Generator](https://github.com/Project-MONAI/tutorials/blob/main/auto3dseg/notebooks/msd_datalist_generator.ipynb).
Let us know if you need further assistance.
Best,

**Reply by che85** [reply]
@KumoLiu Hi. Why is there no easy way to run auto3dseg on new data by just providing a list of files and the original work_dir of the training results? The Slicer MONAIAuto3DSeg module only supports running for a single model (no ensemble is supported). If ensemble is needed, developers recommend using auto3dseg API. Now I am learning, that there is no easy way to run inference for new unseen data that was not part of the original data. 

How is auto3dseg intended to be used post-training if these functionalities are not provided? Is the sole purpose to train and validate a model and then move on?
